File: Leetcode-75/1926-Nearest-Exit-from-Entrance/nearest-exit.py
# Problem #1926 - NEAREST EXIT FROM ENTRANCE IN MAZE

# You are given an m x n matrix maze (0-indexed) with empty cells (represented as '.') and walls (represented as '+'). 
# You are also given the entrance of the maze, where entrance = [entrancerow, entrancecol] denotes the row and column of the cell you are initially standing at.

# In one step, you can move one cell up, down, left, or right. 
# You cannot step into a cell with a wall, and you cannot step outside the maze. 
# Your goal is to find the nearest exit from the entrance. 
# An exit is defined as an empty cell that is at the border of the maze. 
# The entrance does not count as an exit.

# Return the number of steps in the shortest path from the entrance to the nearest exit, or -1 if no such path exists.

#---------------------------------------------------------------------------------------------------------------------------

# Immediately thing BREADTH FIRST SEARCH --> SHORTEST PATH (aka nearest exit)
    # Utilize a QUEUE --> FIFO order
    # Track the cells we have visited
    # Add next coordinates to the Queue to process

#---------------------------------------------------------------------------------------------------------------------------
from collections import deque

def nearestExit(maze: list[list[str]], entrance: list[int]) -> int:
    directions = [(-1,0), (1,0), (0,-1), (0,1)] # Represents four possible movement directions (up, down, left, right)
    entrance = (entrance[0], entrance[1]) # Convert the entrance to a tuple
    queue = deque([entrance]) # Add entrance to the queue for our BFS
    visited = set([entrance]) # Used to keep track of cells visited without duplicates
    step = 0 # Initialized to track number of steps taken --> Return value if exit found

    # Our BFS loop --> Runs until the queue is empty
    # Each iteration represents attempting one step in every direction
    while queue:
        stepLength = len(queue)
        
        for _ in range(stepLength):
            # For each node, we check if it is located on the perimeter of the matrix
            node = queue.popleft()

            if (node[0] == 0 or node[0] == len(maze) - 1 or node[1] == 0 or node[1] == len(maze[0]) - 1) and node != entrance:
                return step # Return current count for shortest answer
            
            # For each node, we check all four directions of movement possible
            for dir in directions:
                newDir = (node[0] + dir[0], node[1] + dir[1]) # Add each direction to the original coordinates (Movement to new space)
                # Here we check threw things:
                    # The cell has not been already visited
                    # The new coordinates are within the bounds of the maze
                    # The new cell is an open path (Represent by the '.' element)
                if newDir not in visited and 0 <= newDir[0] < len(maze) and 0 <= newDir[1] < len(maze[0]) and maze[newDir[0]][newDir[1]] == '.':
                    # If yes to all, add cell to the visited set and add to queue for further exploration
                    visited.add(newDir)
                    queue.append(newDir)

        # After processing all nodes per level, increase step counter to indicate one more step into the maze completed
        step += 1
    
    # If no exit is found through loop, we exit and return -1
    return -1

maze1 = [
    ["+","+",".","+"],
    [".",".",".","+"],
    ["+","+","+","."]
]

maze2 = [
    ["+", ".", "+", "+", "+", "+", "+"],
    ["+", ".", "+", ".", ".", ".", "+"],
    ["+", ".", "+", ".", "+", ".", "+"],
    ["+", ".", ".", ".", ".", ".", "+"],
    ["+", "+", "+", "+", ".", "+", "."],
]

print(nearestExit(maze1, [1,2])) # Expect 1
print(nearestExit(maze2, [0,1])) # Expect 7

File: Leetcode-75/1926-Nearest-Exit-from-Entrance/nearest-exit.ts
// Problem #1926 - NEAREST EXIT FROM ENTRANCE IN MAZE

// BREADTH FIRST SEARCH --> BEST SOLUTION TO FIND THE SHORTEST PATH (AKA CLOSEST EXIT)
function nearestExit(maze: string[][], entrance: number[]): number {
    const dirs = [
        [-1, 0], // Up
        [1, 0], // Down
        [0, -1], // Left
        [0, 1], // Right
    ]; // Four directions we can move from each node

    const queue: [number, number][] = [[entrance[0], entrance[1]]]; // Initialize queue with entrance values

    // Build an array of arrays to track visited cells --> Building as such allows us to store booleans, not processing integers throughout
    const visited = Array.from(
        { length: maze.length }, // First argument is the length of the new array --> Same as total rows in the matrix
        () => Array(maze[0].length).fill(false) // Second argument is how to fill the new array --> In this case we do so by building new arrays for each column of the matrix (length of maze[0])
    );

    const [entranceX, entranceY] = entrance; // Destructure to pull out the x, y coordinates of the entrance
    visited[entranceX][entranceY] = true; // Mark entrance as visited in our array

    let step = 0;

    while (queue.length > 0) {
        const stepLength = queue.length;

        for (let i = 0; i < stepLength; i++) {
            const [x, y] = queue.shift()!; // The queue begins with the entrance --> Each iteration utilizes the newX and newY calculated below

            // check that the X or Y coordinates are along an edge and are not the entrance to the maze
            if (
                (x === 0 ||
                    x === maze.length - 1 ||
                    y === 0 ||
                    y === maze[0].length - 1) &&
                (x !== entranceX || y !== entranceY)
            ) {
                return step; // We've found our exit
            }

            // Move in each direction of the current cell
            for (const [dirX, dirY] of dirs) {
                const newX = x + dirX; // Add the X movement to the current cell location
                const newY = y + dirY; // Add the Y movement to the current cell location

                // Check that the new location is within the bounds of the maze, is not a wall (open is "." value), and it has not been already visited
                if (
                    newX >= 0 &&
                    newX < maze.length &&
                    newY >= 0 &&
                    newY < maze[0].length &&
                    maze[newX][newY] === "." &&
                    !visited[newX][newY]
                ) {
                    visited[newX][newY] = true; // Mark the new cell as visited
                    queue.push([newX, newY]); // Push it's coordinates into the queue to be processed
                }
            }
        }
        step += 1;
    }

    return -1;
}

const maze1 = [
    ["+", "+", ".", "+"],
    [".", ".", ".", "+"],
    ["+", "+", "+", "."],
];

const maze2 = [
    ["+", ".", "+", "+", "+", "+", "+"],
    ["+", ".", "+", ".", ".", ".", "+"],
    ["+", ".", "+", ".", "+", ".", "+"],
    ["+", ".", ".", ".", ".", ".", "+"],
    ["+", "+", "+", "+", ".", "+", "."],
];

console.log(nearestExit(maze1, [1, 2])); // Expect 1
console.log(nearestExit(maze2, [0, 1])); // Expect 7
